src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java

```diff
@@ -259,6 +292,32 @@ public Object getIndex(DynamicObject matchData, DynamicObject range, NotProvided
             return createArray(store, length);
         }
 
+        @TruffleBoundary
+        protected static NameEntry findNameEntry(DynamicObject regexp, DynamicObject string) {
+            Regex regex = Layouts.REGEXP.getRegex(regexp);
+            Rope rope = Layouts.SYMBOL.getRope(string);
+            for (Iterator<NameEntry> i = regex.namedBackrefIterator(); i.hasNext();) {
+                final NameEntry e = i.next();
+                if (bytesEqual(rope.getBytes(), rope.byteLength(), e.name, e.nameP, e.nameEnd)) {
+                    return e;
+                }
+            }
+            return null;
+        }
+
+        protected static DynamicObject getRegexp(DynamicObject matchData) {
+            return Layouts.MATCH_DATA.getRegexp(matchData);
+        }
+
+        private static boolean bytesEqual(byte[] bytes, int byteLength, byte[] name, int nameP, int nameEnd) {
+            if (bytes == name && nameP == 0 && byteLength == nameEnd) {
+                return true;
+            } else {
+                String a = new String(name, nameP, nameEnd - nameP, StandardCharsets.UTF_8);
+                String b = new String(bytes, 0, byteLength, StandardCharsets.UTF_8);
+                return a.equals(b);
```

Comment:
Could a simple loop be used to compare bytes here rather than going though encodings?

Reply:
We have `ArrayUtils.memcmp`, which looks like it'd work well here.